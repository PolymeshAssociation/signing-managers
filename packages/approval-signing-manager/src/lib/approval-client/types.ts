/**
 * Signable data as expected by the Vault API
 */
export interface SignRequestBody {
  /**
   * base64 encoded data to sign
   */
  message: string;

  /**
   * The ID of the owner
   */
  ownerId: string;

  /**
   * The client ID
   */
  apiClientId: string;
}

/**
 * Response representing a signature request. The approval process may require a person so the status change may take hours
 */
export interface GetSignatureResponse {
  id: string;
  virtualHash: string;
  apiClientId: string;
  approvalStatus: string;
  message: string;
  signedMessage: string;
  ownerId: string;
  signed: boolean;
}

/**
 * Response returned when retrieving keys held available for signing with
 */
export interface OwnerKeys {
  /**
   * The keys available to sign with
   */
  accounts: KeyRecord[];

  /**
   * The api client record
   */
  apiClientId: string;

  /**
   * The owner of the key
   */
  ownerId: string;

  /**
   * An internal identifier
   */
  _id: string;
}

/**
 * A record representing a private key stored in the signer
 */
export interface KeyRecord {
  /**
   * The network this key is for. Note this is idiosyncratic to the API this was written against, which is multi-chain
   */
  network: string;

  /**
   * The SS58 encoded public key associated to the stored private key
   */
  address: string;

  /**
   * An internal identifier
   */
  _id: string;
}

export interface KeyRecordWithOwner extends KeyRecord {
  /**
   * The owner of the key record
   */
  ownerId: string;
}

export type Headers = { Authorization: string; 'Content-Type': 'application/json' };

export enum ApprovalStatus {
  Pending = 'pending',
  Approved = 'approved',
}
