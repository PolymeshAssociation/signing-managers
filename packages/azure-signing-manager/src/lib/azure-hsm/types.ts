import { KeyProperties, KeyVaultKey } from '@azure/keyvault-keys';

export interface AzureKey {
  /**
   * Internal key ID.
   */
  id: string;
  /**
   * Internal Azure name. This is used to identify the key within the Key Vault
   */
  name: string;

  /**
   * The public key in raw format
   */
  publicKey: Buffer;

  /**
   * Additional details about the key
   */
  properties: KeyProperties;

  /**
   * The Azure key SDK object. Used for interacting with the key
   */
  keyVaultKey: KeyVaultKey;
}

/**
 * Signable data as expected by the Azure Key Vault API
 */
export interface SignRequestPayload {
  /**
   * base64 encoded data to sign
   */
  input: string;
}

export class AzureSignerError extends Error {
  public data?: Record<string, unknown>;

  /**
   * @hidden
   */
  constructor({ message, data }: { message?: string; data?: Record<string, unknown> }) {
    super(message);

    this.data = data;
  }
}
