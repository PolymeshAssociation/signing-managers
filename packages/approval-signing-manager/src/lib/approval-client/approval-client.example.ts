/* istanbul ignore file */

/**
 * Example file demonstrating calls on the API client
 */

import { ApprovalClient } from '.';

const url = 'https://example.com';
const ownerId = 'someOwnerId';
const apiClientId = 'someClientId';
const apiKey = 'someApiSecret';

const main = async () => {
  const client = new ApprovalClient(url, apiClientId, apiKey, 15);

  const createResponse = await client.createWallet(ownerId);
  console.log('createWallet response: ', createResponse);

  const getKeys = await client.fetchKeys();
  console.log('fetchKeys response: ', getKeys);

  const ownerKeys = await client.fetchOwnerKeys(ownerId);
  console.log('fetchOwnerKeys response: ', ownerKeys);

  /** The backing API needs to approve the transaction before this resolves  */
  const signResponse = await client.signData(ownerId, 'Some base64 encoded Polymesh transaction');
  console.log('signData response: ', signResponse);
};

main();
