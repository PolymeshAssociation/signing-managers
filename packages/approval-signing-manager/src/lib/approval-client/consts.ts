/**
 * A common prefix for all routes
 */
export const routePrefix = 'v1/client';

/**
 * The path to the signature collection
 */
export const signatureRoute = 'polymesh/signature';

/**
 * The path to create a wallet
 */
export const createWalletRoute = 'wallets/create';

/**
 * The path to fetch a clients available wallets. Expected to be appended with the clientId
 */
export const getWalletRoute = 'wallets';

/**
 * The API returns key information for other networks. This is used to filter for polymesh specific values
 */
export const meshNetwork = 'polymesh';
