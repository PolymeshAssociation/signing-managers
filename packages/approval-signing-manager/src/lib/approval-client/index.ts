/* istanbul ignore file */

export { ApprovalClient } from './approval-client';
export * from './types';
