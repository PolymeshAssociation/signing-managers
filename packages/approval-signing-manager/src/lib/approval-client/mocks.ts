import { Headers } from './types';

export class MockApprovalClient {
  public fetchKeys = jest.fn();

  public fetchOwnerKeys = jest.fn();

  public signData = jest.fn();

  public url = 'http://example.com';

  public apiClientId = 'mockClientId';

  public getUrl = jest.fn();

  public getSignature = jest.fn();

  public createWallet = jest.fn();

  private headers: Headers = {
    Authorization: 'mockApiKey',
    'Content-Type': 'application/json',
  };

  private readBody = jest.fn();
}

export const mockApprovalClient = new MockApprovalClient();

export function createMockResponse(
  status: number,
  statusText: string,
  response: unknown
): Response {
  return {
    status,
    statusText,
    json: async () => response,
  } as Response;
}
