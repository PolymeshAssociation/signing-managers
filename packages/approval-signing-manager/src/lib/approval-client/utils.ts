/**
 * A helper that will return a Promise that resolves in the number of milliseconds passed
 *
 * @param ms - number os milliseconds to wait until resolving the returned Promise
 */
export const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
