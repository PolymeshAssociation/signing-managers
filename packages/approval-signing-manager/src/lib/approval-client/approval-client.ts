import { HexString } from '@polkadot/util/types';
import fetch from 'cross-fetch';

import {
  createWalletRoute,
  getWalletRoute,
  meshNetwork,
  routePrefix,
  signatureRoute,
} from './consts';
import {
  ApprovalStatus,
  GetSignatureResponse,
  Headers,
  KeyRecord,
  KeyRecordWithOwner,
  OwnerKeys,
  SignRequestBody,
} from './types';
import { sleep } from './utils';

export class ApprovalClient {
  private headers: Headers;
  private pollingInterval: number;

  /**
   * @hidden
   */
  constructor(
    /**
     * The url serving the API
     */
    private readonly url: string,
    /**
     * The client ID to be intended to be authorized for
     */
    private readonly apiClientId: string,
    /**
     * The apiKey to grant authorization
     */
    readonly apiKey: string,
    /**
     * The number of seconds to wait before checking for a signature. Note, each pending transaction will poll this often
     */
    pollingInterval: number
  ) {
    this.headers = {
      Authorization: apiKey,
      'Content-Type': 'application/json',
    };
    this.pollingInterval = pollingInterval * 1000; // convert to milliseconds
  }

  /**
   * Sign data on behalf of a key and return the hex signature
   * @note this method will continue to poll an endpoint until the transaction is approved. It may take a while for the promise to resolve
   *
   * @param ownerId - id of the key that will sign
   * @param message - the hex encoded data that will be signed
   */
  public async signData(ownerId: string, message: string): Promise<HexString> {
    const { headers, apiClientId } = this;

    const requestSignatureUrl = this.getUrl(signatureRoute);
    const payload: SignRequestBody = { message, apiClientId, ownerId };
    const body = JSON.stringify(payload);

    const response = await fetch(requestSignatureUrl, {
      method: 'POST',
      body,
      headers,
    });

    let responseJson = await this.readBody<GetSignatureResponse>(response);

    while (responseJson.approvalStatus === ApprovalStatus.Pending) {
      responseJson = await this.getSignature(responseJson.virtualHash);
      await sleep(this.pollingInterval);
    }

    const { signedMessage, signed } = responseJson;

    if (!signed || !signedMessage) {
      throw new Error(
        'The message was not signed and the request is no longer pending. The transaction has likely been rejected'
      );
    }

    // prepend a byte with value `0x01` to indicate this is an sr25519 signature
    const hexSignature = signedMessage.replace('0x', '');

    return `0x01${hexSignature}`;
  }

  /**
   * @hidden
   *
   * Assemble URL from separate path strings
   */
  private getUrl(...path: string[]) {
    return `${this.url}/${routePrefix}/${path.join('/')}`;
  }

  /**
   * Get the signature information for the given `virtualHash`
   */
  private async getSignature(virtualHash: string): Promise<GetSignatureResponse> {
    const { headers } = this;
    const url = this.getUrl(signatureRoute, virtualHash);

    const response = await fetch(url, { headers });
    return this.readBody<GetSignatureResponse>(response);
  }

  /**
   * Get the set of accounts accessible by the current API key
   */
  public async fetchKeys(): Promise<KeyRecordWithOwner[]> {
    const { headers } = this;
    const url = this.getUrl(getWalletRoute, this.apiClientId);

    const response = await fetch(url, { headers });
    const keys = await this.readBody<OwnerKeys[]>(response);
    console.log('keys', keys);

    return keys.flatMap(({ ownerId, accounts }) =>
      accounts
        .filter(({ network }) => network === meshNetwork)
        .map(account => ({ ...account, ownerId }))
    );
  }

  /**
   * Get the set of accounts belonging to the given `ownerId`
   */
  public async fetchOwnerKeys(ownerId: string): Promise<KeyRecordWithOwner[]> {
    const { headers } = this;
    const url = this.getUrl(getWalletRoute, this.apiClientId, ownerId);

    const response = await fetch(url, { headers });
    const keys = await this.readBody<OwnerKeys>(response);

    return keys.accounts.filter(isMeshNetwork).map(account => ({ ...account, ownerId }));
  }

  /**
   *
   * @param ownerId - The identifier that will be given to the wallet
   * @returns - The created key information
   */
  public async createWallet(ownerId: string): Promise<KeyRecordWithOwner[]> {
    const { headers, apiClientId } = this;
    const url = this.getUrl(createWalletRoute);
    const body = JSON.stringify({ ownerId, apiClientId });
    const response = await fetch(url, {
      method: 'POST',
      body,
      headers,
    });
    const responseBody = await this.readBody<OwnerKeys>(response);

    return responseBody.accounts.filter(isMeshNetwork).map(account => ({ ...account, ownerId }));
  }

  /**
   * Ensures the response was successful, returns the response body as json
   *
   * @note the type param `T` acts as a type assertion for the response, no runtime checks are performed to ensure the response conforms to the expectation
   *
   * @throws if the status code was not in the 2xx range
   */
  private async readBody<T>(res: Response): Promise<T> {
    const { status, statusText } = res;
    const body = await res.json();

    if (status >= 300) {
      const { message } = body;
      throw new Error(
        `Approval client response error: ${status} - ${statusText}. Reason(s): ${message}`
      );
    }

    return body;
  }
}

const isMeshNetwork = ({ network }: KeyRecord) => network === meshNetwork;
