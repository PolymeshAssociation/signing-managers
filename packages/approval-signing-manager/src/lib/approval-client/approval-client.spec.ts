import fetch from 'cross-fetch';

import { ApprovalClient } from './approval-client';
import { createMockResponse } from './mocks';
import { KeyRecord } from './types';

jest.mock('cross-fetch', () => jest.fn());

const fetchStub = fetch as jest.Mock;

describe('ApprovalClient class', () => {
  const url = 'http://localhost';
  const ownerId = 'someOwnerId';
  const apiClientId = 'someId';
  const token = 'someToken';

  const accountResponse: KeyRecord[] = [
    {
      address: '0x01000',
      network: 'polymesh',
      _id: '1',
    },
    {
      address: '0x02000',
      network: 'polymesh',
      _id: '2',
    },
    {
      address: '0x3000',
      network: 'polymesh',
      _id: '3',
    },
  ];

  const otherNetworkAccounts: KeyRecord[] = [{ address: '0x04000', network: 'ethereum', _id: '4' }];

  let client: ApprovalClient;

  beforeEach(() => {
    client = new ApprovalClient(url, apiClientId, token, 1);
  });

  describe('method: fetchKeys', () => {
    it('should return all accessible keys', async () => {
      fetchStub.mockResolvedValue(
        createMockResponse(200, 'Ok', [
          {
            apiClientId,
            ownerId,
            accounts: [...accountResponse, ...otherNetworkAccounts],
          },
        ])
      );

      const result = await client.fetchKeys();
      expect(result).toMatchSnapshot();
    });

    it('should throw any errors returned by the Approver API', () => {
      fetchStub.mockResolvedValue(
        createMockResponse(400, 'Bad Request', {
          message: 'Something went wrong',
        })
      );
      return expect(client.fetchKeys()).rejects.toThrow(
        `Approval client response error: 400 - Bad Request. Reason(s): Something went wrong`
      );
    });
  });

  describe('method: fetchOwnerKeys', () => {
    it('should return all keys under the passed owner', async () => {
      fetchStub.mockResolvedValueOnce(
        createMockResponse(200, 'Ok', {
          accounts: [accountResponse[0], otherNetworkAccounts[0]],
        })
      );

      const result = await client.fetchOwnerKeys(ownerId);
      expect(result).toMatchSnapshot();
    });
  });

  describe('method: signData', () => {
    it('should return signed data', async () => {
      fetchStub
        .mockResolvedValueOnce(
          createMockResponse(200, 'Ok', {
            approvalStatus: 'pending',
            virtualHash: '123',
            signed: false,
          })
        )
        .mockResolvedValueOnce(
          createMockResponse(200, 'Ok', {
            approvalStatus: 'approved',
            virtualHash: '123',
            signed: true,
            signedMessage: '0xabc',
          })
        );

      const result = await client.signData('Alice', 'adba32==');

      expect(result).toBe('0x01abc');
    });

    it('should throw if the transaction is rejected', () => {
      fetchStub
        .mockResolvedValueOnce(
          createMockResponse(200, 'Ok', {
            approvalStatus: 'pending',
            virtualHash: '123',
            signed: false,
          })
        )
        .mockResolvedValueOnce(
          createMockResponse(200, 'Ok', {
            approvalStatus: 'rejected',
            virtualHash: '123',
            signed: false,
          })
        );

      return expect(client.signData('Alice', 'adba32==')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw any errors returned by the Approver API', () => {
      fetchStub.mockResolvedValue(
        createMockResponse(400, 'Bad Request', {
          message: 'THE END IS NIGH',
        })
      );

      return expect(client.fetchKeys()).rejects.toThrow(
        `Approval client response error: 400 - Bad Request. Reason(s): THE END IS NIGH`
      );
    });
  });

  describe('method: createWallet', () => {
    it('should call the endpoint to create a wallet', async () => {
      fetchStub.mockResolvedValue(
        createMockResponse(200, 'Ok', {
          apiClientId,
          ownerId,
          accounts: [
            {
              network: 'eth',
              token: 'eth',
              address: '0x3f9e4839005593f9fEb30FC7f87a777396c36106',
              _id: '62fc2a617cbb5b9ecbaff97d',
            },
            {
              network: 'polymesh',
              token: 'polyx',
              address: '5GdqFHLLxZt2u6HokFLBnyopDnGsS2sCBod4vUZTQ7YYx',
              _id: '62fc2a617cbb5b9ecbaff97f',
            },
          ],
        })
      );

      const result = await client.createWallet(ownerId);

      expect(result).toMatchSnapshot();
    });
  });
});
