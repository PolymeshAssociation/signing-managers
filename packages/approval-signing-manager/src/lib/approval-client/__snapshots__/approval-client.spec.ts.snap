// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApprovalClient class method: createWallet should call the endpoint to create a wallet 1`] = `
Array [
  Object {
    "_id": "62fc2a617cbb5b9ecbaff97f",
    "address": "5GdqFHLLxZt2u6HokFLBnyopDnGsS2sCBod4vUZTQ7YYx",
    "network": "polymesh",
    "ownerId": "someOwnerId",
    "token": "polyx",
  },
]
`;

exports[`ApprovalClient class method: fetchKeys should return all accessible keys 1`] = `
Array [
  Object {
    "_id": "1",
    "address": "0x01000",
    "network": "polymesh",
    "ownerId": "someOwnerId",
  },
  Object {
    "_id": "2",
    "address": "0x02000",
    "network": "polymesh",
    "ownerId": "someOwnerId",
  },
  Object {
    "_id": "3",
    "address": "0x3000",
    "network": "polymesh",
    "ownerId": "someOwnerId",
  },
]
`;

exports[`ApprovalClient class method: fetchOwnerKeys should return all keys under the passed owner 1`] = `
Array [
  Object {
    "_id": "1",
    "address": "0x01000",
    "network": "polymesh",
    "ownerId": "someOwnerId",
  },
]
`;

exports[`ApprovalClient class method: signData should throw if the transaction is rejected 1`] = `"The message was not signed and the request is no longer pending. The transaction has likely been rejected"`;
